### ISSUE TITLE
accept() on listening socket without changing netns

### ISSUE THREAD

**Issue by peterjin-org** [vi/netns_tcp_bridge#1]
Wouldn't it be more efficient to just create the listening socket in one of the network namespaces and then accept() and connect() in the same thread? It is not necessary to change the network namespace of the process to use a listening socket, in particular, if a process creates a socket, then it is bound to the network namespace in which it is created, and this binding is retained on all socket calls made on it even if it is moved to a different network namespace.

**Comment by peterjin-org** [comment]
In other words, you would create the listening socket in one netns, then change to the "connection" netns, and then run a regular TCP relay server.

**Comment by vi** [comment]
Indeed, that should simplify the code a bit and reduce new forwarded connections latency by a couple of syscalls. And after the initial setup it could be strictly single-threaded with one process.

On the other hand, current design may serve as a template to make some load balancer or other trickier tool.

**Comment by peterjin-org** [comment]
I usually do that using LD_PRELOAD on the server application to emulate accept() using SCM_RIGHTS (see the socketbox repo in my account), which will work on any server application without modification of the server code itself.

**Comment by vi** [comment]
> LD_PRELOAD
> accept
> SCM_RIGHTS

Looks like another valid approach. Maybe socketboxed `socat` may serve as an alternative for `netns_tcp_bridge`? Note sure where to specify the netns file though. Also not sure how to manage the UNIX socket for `SCM_RIGHTS` - brief reading of socketbox's README has not revealed it. `netns_tcp_bridge` does not require creating any entities like filesystem-based sockets - you terminate the process and forwarding gets stopped without any leftover cruft (maybe just the lingering socket).

The LD_PRELOAD approach will fail for server apps that are built with static libc (i.e. many Goland or Rust apps).

---

Is it a just a discussion issue or you expect me to do some changes to `netns_tcp_bridge` to suit your use case?

**Comment by peterjin-org** [comment]
Yeah, sort of. To specify a different network namespace for socketbox you would have to run the socketbox server in the namespace where you would like the listening socket to be, and then run your application in a different network namespace from socketbox. I sadly haven't updated the socketbox readme since I first wrote it in late-2020; more information about socketbox can be found on my website at https://website.peterjin.org/wiki/Socketbox

Socketbox-preload automatically removes the listening SOCK_DGRAM socket if it already exists. I was thinking of Go bindings for Socketbox at one point, but for the applications that I run on my servers, they're not necessary at this point.

I brought up socketbox here since it's quite funny that you and me have come to the same problem (forwarding connections between network namespaces without using veth pairs) and ended up with very similar solutions :) I actually came from your Reddit post, but I don't have a Reddit account. I have also posted a number of other network namespace ideas to think about at https://website.peterjin.org/wiki/Notes_about_namespaces#Network_namespaces

I only looked at your code for a bit, but I think that netns_tcp_bridge may actually be able to support the Socketbox "B" protocol without too much modification since they both use SOCK_SEQPACKET Unix domain sockets with SCM_RIGHTS to send accepted connections.

Are you familiar with the slirp4netns project?

**Comment by vi** [comment]
> Yeah, sort of.

Unclear what this sentence does reply to.

> Socketbox-preload automatically removes the listening SOCK_DGRAM socket if it already exists.

So user needs to invent some file path for the socket and a shared filesystem between the host and container? `netns_tcp_bridge` uses `socketpair(2)`, saving user from hassle of creating a filename.

> I was thinking of Go bindings for Socketbox at one point, but for the applications that I run on my servers, they're not necessary at this point.

Does Go offer something similar to LD_PRELOAD? Or something similar (i.e. "monkey patching") at source code level?

> https://website.peterjin.org/wiki/Notes_about_namespaces

Thanks for sharing it. In progress of reading.

> ... netns_tcp_bridge ...  to support the Socketbox "B" protocol ...

Opening UNIX socket by path may be out of scope for `netns_tcp_bridge`.
And why implement socketbox mode in netns_tcp_bridge if there is already socketbox-relay?

> Are you familiar with the slirp4netns project?

No. I see it is mentioned in the wiki.

It looks like a tool to use with user namespaces, which I currently do not use (as they expose large kernel code attack surface, allowing more local root exploits).

I remember relying on slirp_fullbolt and pppd to make a makeshift VPN on openvz VPS. When I use qemu-system, it uses slirp inside, as far as I know.
